include/swift/AST/Module.h

```diff
@@ -744,7 +748,7 @@ class FileUnit : public DeclContext {
   // Only allow allocation of FileUnits using the allocator in ASTContext
   // or by doing a placement new.
   void *operator new(size_t Bytes, ASTContext &C,
-                     unsigned Alignment = alignof(FileUnit));
+                     unsigned Alignment = alignof(FileUnit&));
```

Comment:
The following modification of your example works:
```
struct Foo {
  bool b;
};

 static_assert(alignof(Foo&) == alignof(bool), "aligning reference rather than Foo");
```